krm-functions/lib/fn-runtime/downstream_inventory.go

```diff
@@ -0,0 +1,129 @@
+	return &downstreamInventory{
+		namedResources: map[string]*downstreamResources{},
+	}
+}
+
+type downstreamInventory struct {
+	namedResources map[string]*downstreamResources
+}
+
+type downstreamResources struct {
+	resources    map[corev1.ObjectReference]*downstreamInventoryCtx
+	forObj       *fn.KubeObject
+	forCondition kptv1.Condition
+}
+
+type downstreamInventoryCtx struct {
+	condition *kptv1.Condition
+	obj       *fn.KubeObject
+}
+
+func (r *downstreamInventory) AddCondition(name string, resource *corev1.ObjectReference, c *kptv1.Condition) {
+	if _, ok := r.namedResources[name]; !ok {
+		r.namedResources[name] = &downstreamResources{
+			resources: map[corev1.ObjectReference]*downstreamInventoryCtx{},
+		}
+	}
+	if r.namedResources[name].resources[*resource] == nil {
+		r.namedResources[name].resources[*resource] = &downstreamInventoryCtx{}
+	}
+	r.namedResources[name].resources[*resource].condition = c
+}
+
+func (r *downstreamInventory) AddResource(name string, resource *corev1.ObjectReference, o *fn.KubeObject) {
+	if _, ok := r.namedResources[name]; !ok {
+		r.namedResources[name] = &downstreamResources{
+			resources: map[corev1.ObjectReference]*downstreamInventoryCtx{},
+		}
+	}
+	if r.namedResources[name].resources[*resource] == nil {
+		r.namedResources[name].resources[*resource] = &downstreamInventoryCtx{}
+	}
+	r.namedResources[name].resources[*resource].obj = o
+}
+
+func (r *downstreamInventory) AddForCondition(name string, c *kptv1.Condition) {
+	if _, ok := r.namedResources[name]; !ok {
+		r.namedResources[name] = &downstreamResources{
+			resources: map[corev1.ObjectReference]*downstreamInventoryCtx{},
+		}
+	}
+	r.namedResources[name].forCondition = *c
+}
+
+func (r *downstreamInventory) AddForResource(name string, o *fn.KubeObject) {
+	if _, ok := r.namedResources[name]; !ok {
+		r.namedResources[name] = &downstreamResources{
+			resources: map[corev1.ObjectReference]*downstreamInventoryCtx{},
+		}
+	}
+	r.namedResources[name].forObj = o
+}
+
+func (r *downstreamInventory) GetForConditionStatus(name string) bool {
+	c, ok := r.namedResources[name]
+	if !ok {
+		// this should not happen
+		return false
+	}
+	return c.forCondition.Status == kptv1.ConditionFalse
+}
+
+func (r *downstreamInventory) GetReadinessStatus() map[string]*ReadyCtx {
+	readyMap := map[string]*ReadyCtx{}
+	for name, ref := range r.namedResources {
+		readyMap[name] = &ReadyCtx{
+			ForCondition: ref.forCondition,
+			Objs:         map[corev1.ObjectReference]fn.KubeObject{},
+		}
+		if ref.forObj != nil {
+			readyMap[name].ForObj = ref.forObj
+		}
+
+		// if no child resources exist we return empty
+		if len(ref.resources) == 0 {
+			readyMap[name].Ready = false
+			return readyMap
+		}
+
+		readyMap[name].Ready = true
+		for objRef, invCtx := range ref.resources {
+			// not all resources have a condition set e.g. interface
+			//
+			if invCtx.condition == nil || (invCtx.condition != nil && invCtx.condition.Status == kptv1.ConditionFalse) {
```

Comment:
For 2, If `condition == nil`, then we consider `ForObj` not ready. What the code now does is break and not continue since it already encountered a failure for that `namedresource`.

Reply:
On 1 it is an or so I believe we need both as we want to check the combination.

Reply:
if invCtx is nil the second statement will not be evaluated as the output us True anyway. Hence we do not need to check for nil in the second condition. If the second condition is executed intCtx.condition will be not nil.

Reply:
agreed, sorry jet lag. will update it